src/UnionFind.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module implements a persistent union-find data structure. This
+-- structure reprensents a map from equivalence classes of keys to values. The
+-- equivalence relation is irrelevant. The structure primarily provides two
+-- operations:
+--
+-- - @union@ takes two keys and merges their equivalence classes, returning a
+--   new structure. Because both keys might be associated to a value, this
+--   operation must also be told how to reconcile both values into one.
+--
+-- - @find@ takes a key and returns the value associated to the key's
+--   equivalence class. In order for the structure to remain efficient, @find@
+--   also optimises it for future calls and therefore returns a new structure.
+
+module UnionFind where
+
+import Data.Default (Default, def)
+import Data.Either (partitionEithers)
+import Data.Function ((&))
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust)
+import Prettyprinter
+
+data UnionFindCell key value
+  = ChildOf key
+  | Ancestor
+  | AncestorWith value
```

Comment:
If I'm understanding correctly, we need this because of the above issue, where we might have variables that are equal but unassigned. In that case, we can't just disallow `Ancestor` can we?

Reply:
That is correct.

Reply:
Note: I merged these two together to simplify other pieces of code. This now looks like:
https://github.com/tweag/pirouette/blob/531cb2e35b7e78e932e9230b8f95e46a98840415/src/UnionFind.hs#L29-L31

Reply:
Seems better to me